src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,279 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using CodeAction = Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Resolves a code action by filling out its Edit or Command property.
+    /// </summary>
+    [ExportLspMethod(MSLSPMethods.TextDocumentCodeActionResolveName), Shared]
+    internal class CodeActionResolveHandler : AbstractRequestHandler<LSP.VSCodeAction, LSP.VSCodeAction>
+    {
+        private readonly ICodeFixService _codeFixService;
+        private readonly ICodeRefactoringService _codeRefactoringService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionResolveHandler(
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            ILspSolutionProvider solutionProvider)
+            : base(solutionProvider)
+        {
+            _codeFixService = codeFixService;
+            _codeRefactoringService = codeRefactoringService;
+        }
+
+        public override async Task<LSP.VSCodeAction> HandleRequestAsync(
+            LSP.VSCodeAction codeAction,
+            LSP.ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            CodeActionResolveData data;
+            if (codeAction.Data is CodeActionResolveData codeActionResolveData)
```

Comment:
this was due to liveshare previously deserializing things for us, I believe it should only ever be jtoken now that it's coming through the LSP client.

Reply:
I think we need to keep in the first check for the code action resolver tests. I can make a comment to note that. I'm not sure if there's a workaround?

Reply:
I think in the tests we can serialize it to string before putting it in the data field.  would be better than keeping this check around I think